scripts/sanitize_metadata.py

```diff
@@ -1,32 +1,282 @@
+
+    # Create a default mapping of location fields to missing values and update
+    # these from the values in the location string.
+    locations = {field: "?" for field in location_fields}
+    locations.update(dict(zip(location_fields, values)))
+
+    return locations
+
+
+def resolve_duplicates(metadata, strain_field, error_on_duplicates=False):
+    """Resolve duplicate records for a given strain field and return a deduplicated
+    data frame. This approach chooses the record with the most recent database
+    accession, if accession fields exist, or the first record for a given strain
+    name. Optionally, raises an error when duplicates are detected, reporting
+    the list of those duplicate records.
+
+    Parameters
+    ----------
+    metadata : pandas.DataFrame
+        Metadata table that may or may not have duplicate records by the given strain field.
+
+    strain_field : string
+        Name of the metadata field corresponding to the strain name and which is used to identify duplicates.
+
+    error_on_duplicates : boolean
+        Whether to throw an error on detection of duplicates or not.
+
+    Returns
+    -------
+    pandas.DataFrame :
+        Metadata with no duplicate records by the given strain field.
+
+    Raises
+    ------
+    ValueError
+        If duplicates are detected and `error_on_duplicates` is `True`.
+    """
+    # Create a list of duplicate records by strain name.
+    duplicates = metadata.loc[
+        metadata.duplicated(strain_field),
+        strain_field
+    ].values
+
+    if len(duplicates) == 0:
+        # No duplicates, so return the original metadata.
+        return metadata
+
+    if error_on_duplicates:
+        # Duplicates and error requested on duplicates, so throw an exception
+        # with the list of duplicate strains.
+        raise ValueError(", ".join(duplicates))
+
+    # Try to resolve the duplicates. First, check for standard accession fields.
+    accession_fields = [
+        field
+        for field in ACCESSION_FIELDS
+        if field in metadata.columns
+    ]
+
+    # If any of these fields exists, sort by strain name and accessions in
+    # ascending order and take the last record. Otherwise, sort and group by
+    # strain name and take the last record from each group. It is possible for
+    # metadata to contain fields for multiple database accessions and for these
+    # fields to be incomplete for some databases. By sorting across all fields,
+    # we use information from all available accession fields. If all fields
+    # contain missing values (e.g., "?"), we end up returning the last record
+    # for a given strain as a reasonable default.
+    sort_fields = [strain_field]
+    if len(accession_fields) > 0:
+        sort_fields.extend(accession_fields)
+
+    # Return the last record from each group after sorting.
+    return metadata.sort_values(sort_fields).groupby(strain_field, sort=False).last().reset_index()
```

Comment:
Maybe this is a more transparent way of doing this:
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html

Reply:
This implementation is a little more complicated because it tries to resolve duplicates by preferring records with the most recent accession from GISAID or GenBank instead of taking the first/last record by default. This prefers the latest metadata record, although the bigger issue is preferring the latest sequence record and we can't do that unless GISAID includes the record's accession with the sequence header.

I just expanded the inline comments around this code, though, to clarify the context of why it's sorting and then grouping the way it is.

Reply:
I didn't notice the `keep` argument for `drop_duplicates` before. That does make the code clearer and a little faster, too. Thanks, @rneher!